x-pack/plugins/endpoint/server/routes/resolver.ts

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  IRouter,
+  ResponseError,
+  Logger,
+  RequestHandler,
+  IScopedClusterClient,
+  KibanaResponseFactory,
+  IKibanaResponse,
+} from 'kibana/server';
+import { SearchResponse } from 'elasticsearch';
+import { schema } from '@kbn/config-schema';
+
+import { EndpointAppContext, Total, JSONish } from '../types';
+import {
+  getESChildrenQuery,
+  getESNodeQuery,
+  getESChildrenCountQuery,
+  getESNodeCountQuery,
+  PaginationInfo,
+} from '../services/resolver/query_builder';
+import { BaseSearchHandler } from '../services/resolver/search_handler';
+import { ResolverData } from '../../common/types';
+import { EntityParseError, CountQueryInfo } from '../services/resolver/common';
+import { ChildrenSearchHandler } from '../services/resolver/children_search_handler';
+import { SingleNodeHandler } from '../services/resolver/single_node_handler';
+
+interface ResolverRequestParams {
+  page_index?: number;
+  page_size?: number;
+  entity_id: string;
+}
+
+type SearchQueryBuilder = (
+  endpointContext: EndpointAppContext,
+  entityID: string,
+  paginationInfo: PaginationInfo
+) => Promise<JSONish>;
+
+type CountQueryBuilder = (entityID: string) => CountQueryInfo;
+
+type SearchHandlerCreator = (
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) => BaseSearchHandler;
+
+function createChildrenSearchHandler(
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) {
+  return new ChildrenSearchHandler(client, endpointContext, pageInfo, countQuery, entityID);
+}
+
+function createSingleNodeHandler(
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) {
+  return new SingleNodeHandler(client, endpointContext, pageInfo, countQuery, entityID);
+}
+
+function getReqHandler(
+  searchQueryBuilder: SearchQueryBuilder,
+  countQueryBuilder: CountQueryBuilder,
+  debugQueryString: string,
+  notFoundString: string,
+  log: Logger,
+  endpointAppContext: EndpointAppContext,
+  handlerCreator: SearchHandlerCreator
+): RequestHandler<unknown, ResolverRequestParams> {
+  return async (context, req, res) => {
+    const entityID = req.query.entity_id;
+    log.debug(`entity_id: ${entityID}`);
+    const paginationInfo = {
+      page: req.query?.page_index,
+      pageSize: req.query?.page_size,
+    } as PaginationInfo;
+    try {
+      const query = await searchQueryBuilder(endpointAppContext, entityID, paginationInfo);
```

Comment:
I think this would be easier to read if generating the query was separate from adding the pagination and index parameters to it.  There are 4 different cases for the searchQueryBuilder (legacy ESChildren, ESChildren, legacy ESNode, and ESNode) but they all use the same `buildSearchBody(context, query, paginationInfo, index)` function call to add pagination parameters after creating the core of the query.  Tracing through a relatively deep call stack like getReqHandler -> searchQueryBuilder -> getESChildrenQuery -> buildSearchBody -> getPagination adds confusion imo.

I propose returning the query and index in one function call here, like `const { index, query } = getESChildrenCountQuery(entityID);` and then using that result in the next line as `const enrichedQuery = buildSearchBody(context, query, paginationInfo, index);`.  The advantage here is that it's clear that the pagination parameters are created the same way in all 4 cases and the call stack readers have to trace through isn't as deep.  I'd also like to rename `buildSearchBody` to something more descriptive like 'addQueryPaginationParameters`.